lib/queuePopulator/LogReader.js

```diff
@@ -316,21 +318,87 @@ class LogReader {
                 this._processLogEntry(batchState, record, entry);
             });
             this._unsetEntryBatch();
-        });
-        logRes.log.on('error', err => {
-            this.log.error('error fetching entries from log',
-                { method: 'LogReader._processPrepareEntries',
-                    error: err });
-            return done(err);
-        });
-        logRes.log.on('end', () => {
-            this.log.debug('ending record stream');
-            return done();
-        });
-        logRes.log.on('info', info => {
-            this.log.debug('received record stream "info" event', { info });
-            logRes.info = info;
-        });
+
+            // Pause tailable streams when reaching the record batch
+            // limit, as those streams do not emit 'end' events but
+            // continuously send 'data' events as they come.
+            if (logRes.tailable &&
+                logStats.nbLogRecordsRead >= batchState.maxRead) {
+                logger.debug('ending batch', {
+                    method: 'LogReader._processPrepareEntries',
+                    reason: 'limit on number of read records reached',
+                    readRecords: logStats.nbLogRecordsRead,
+                });
+                shipBatchCb();
+            }
+        };
+        const errorEventHandler = err => {
+            logger.error('error fetching entries from log', {
+                method: 'LogReader._processPrepareEntries',
+                error: err,
+            });
+            return shipBatchCb(err);
+        };
+        const endEventHandler = () => {
+            logger.debug('ending record stream', {
+                method: 'LogReader._processPrepareEntries',
+            });
+            return shipBatchCb();
+        };
+        logRes.log.on('data', dataEventHandler);
+        logRes.log.on('error', errorEventHandler);
+        logRes.log.on('end', endEventHandler);
```

Comment:
maybe an event might emit before `shipBatchCb` is set

Reply:
I don't think it's a possibility, since new event handlers should not get events already queued for delivery, I think, only the newer events will be dispatched, which can occur only after we return from the function `_processPrepareEntries()`